internal/aws/awsutil/conn.go

```diff
@@ -203,29 +203,51 @@ func GetAWSConfigSession(logger *zap.Logger, cn ConnAttr, cfg *AWSSessionSetting
 		CredentialsChainVerboseErrors: aws.Bool(true),
 	}
-	// do not overwrite for sts assume role
-	if cfg.RoleARN == "" && len(override.GetCredentialsChainOverride().GetCredentialsChain()) > 0 {
+	// if role arn is used do not try other creds
+	if cfg.RoleARN == "" {
 		config.Credentials = credentials.NewCredentials(&credentials.ChainProvider{
-			Providers: customCredentialProvider(cfg, config),
+			Providers: customCredentialProvider(cfg, config, s),
 		})
 	}
 	return config, s, nil
 }
 
-func customCredentialProvider(cfg *AWSSessionSettings, config *aws.Config) []credentials.Provider {
-	defaultCredProviders := defaults.CredProviders(config, defaults.Handlers())
+// order
+// 1. shared profile
+// 2. env var
+// 3. shared creds file
+// 4. ec2 creds
+// 5. override creds providers
+// explicit keys do not make sense in this context since contrib does not take in explicit keys
+// and cwa does not provide explicit keys to config but come from these methods of getting creds
+// @TODO make this more upstream friendly working in the future
+func customCredentialProvider(cfg *AWSSessionSettings, config *aws.Config, sess *session.Session) []credentials.Provider {
 	overrideCredProviders := override.GetCredentialsChainOverride().GetCredentialsChain()
 	credProviders := make([]credentials.Provider, 0)
-	// if is for differently configured shared creds file location
-	// else if is for diff profile but no change in creds file ex run in containers
-	if cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
+	// this is to keep ordering where if there is a profile we want to try those creds first
+	if cfg.Profile != "" && cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
 		for _, file := range cfg.SharedCredentialsFile {
-			credProviders = append(credProviders, &credentials.SharedCredentialsProvider{Filename: file, Profile: cfg.Profile})
+			sharedCredentialsProvider := credentials.SharedCredentialsProvider{Filename: file, Profile: cfg.Profile}
+			credProviders = append(credProviders, &sharedCredentialsProvider)
+			credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{sharedCredentialsProvider: &sharedCredentialsProvider})
 		}
 	} else if cfg.Profile != "" {
-		credProviders = append(credProviders, &credentials.SharedCredentialsProvider{Filename: "", Profile: cfg.Profile})
+		sharedCredentialsProvider := credentials.SharedCredentialsProvider{Filename: "", Profile: cfg.Profile}
+		credProviders = append(credProviders, &sharedCredentialsProvider)
+		credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{sharedCredentialsProvider: &sharedCredentialsProvider})
 	}
-	credProviders = append(credProviders, defaultCredProviders...)
+	credProviders = append(credProviders, &credentials.EnvProvider{})
+	if cfg.Profile == "" && cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
+		for _, file := range cfg.SharedCredentialsFile {
```

Comment:
Lines 241-244 seem to be the same as lines 229-233 so I initially thought we could drop the `cfg.Profile != ""` check on line 228 and thatll reduce the code, but doing that would change the creds order wouldnt it?

Reply:
Confused on this as well. Do we need the `cfg.Profile != ""` check? It looks empty or non-empty, we are setting the credential to whatever is in the file, and if empty, SDK will just fallback to [default](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#SharedCredentialsProvider)?

Reply:
This is because of the creds order. We want when a profile is given to always be before a profile not being given in the cred provider. This is based on the comments. Take for instance we pass in a profile but no file.